virtcontainers/device/manager/manager.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2017-2018 Intel Corporation
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package manager
+
+import (
+	"github.com/sirupsen/logrus"
+
+	"github.com/kata-containers/runtime/virtcontainers/device/api"
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+	"github.com/kata-containers/runtime/virtcontainers/device/drivers"
+)
+
+const (
+	// VirtioBlock indicates block driver is virtio-blk based
+	VirtioBlock string = "virtio-blk"
+	// VirtioSCSI indicates block driver is virtio-scsi based
+	VirtioSCSI string = "virtio-scsi"
+)
+
+type deviceManager struct {
+	blockDriver string
+}
+
+func deviceLogger() *logrus.Entry {
+	return api.DeviceLogger().WithField("subsystem", "device")
+}
+
+// CreateDevice creates one device based on DeviceInfo
+func (dm *deviceManager) CreateDevice(devInfo config.DeviceInfo) api.Device {
+	path := devInfo.HostPath
+
+	if isVFIO(path) {
+		return drivers.NewVFIODevice(devInfo)
+	} else if isBlock(devInfo) {
+		if devInfo.DriverOptions == nil {
+			devInfo.DriverOptions = make(map[string]string)
+		}
+		devInfo.DriverOptions["block-driver"] = dm.blockDriver
+		return drivers.NewBlockDevice(devInfo)
+	} else {
+		deviceLogger().WithField("device", path).Info("Device has not been passed to the container")
+		return drivers.NewGenericDevice(devInfo)
+	}
+}
+
+// NewDevices creates bundles of devices based on array of DeviceInfo
+func (dm *deviceManager) NewDevices(devInfos []config.DeviceInfo) ([]api.Device, error) {
+	var devices []api.Device
+
+	for _, devInfo := range devInfos {
+		hostPath, err := config.GetHostPathFunc(devInfo)
```

Comment:
Please move this inside dm.CreateDevice() so that the two dm APIs accept the same DeviceInfo structure.

Reply:
There's one thing I'm not sure here, `config.GetHostPathFunc` would give a different hostPath, and `NewDevices()` will modify `devInfo.HostPath`, and `CreateDevice()` won't modify devInfo . This is a difference between `NewDevices()` and `CreateDevice()` (modify hostPath or not).

Are these two APIs designed so? Because I'm not certain on this so I didn't modify it.
Do you think it's safe to move it inside `dm.CreateDevice()`?

Reply:
`CreateDevice()` is only called inside `NewDevices()`. So I think it is safe and we should let them share a similar semantics w.r.t. deviceinfo, unless you have other plans to use `CreateDevice()` differently in your storage hotplug PR.

Reply:
Oh, you're right! And I think then we only need to expose NewDevices and don't need to expose `CreateDevice` now!
Great suggestion! I will modify this part.

Reply:
@bergwolf Updated